### ISSUE TITLE
Improve testing / continuous integration

### ISSUE THREAD

**Issue by scls19fr** [JuliaIO/LibSerialPort.jl#16]
Hello,

it will be nice to improve testing of this library.

You could:

- Add a `test/runtests.jl` file like most Julia package does 

with

```julia
using LibSerialPort
using Test

include("test-low-level-api.jl")
include("test-high-level-api.jl")
```

- Use `@testset` macro https://docs.julialang.org/en/v0.7.0/stdlib/Test/#Working-with-Test-Sets-1

- Enable continuous integration
  - With Travis https://travis-ci.com for Linux and Mac testing (a `.travis.yml` file is required)
  - With Appveyor https://www.appveyor.com/ for Windows testing (a `appveyor.yml` file is required)

It will at least insure that this library can be built.

- Enable code coverage (for perfectionists)

You should probably create a dummy package using the new package manager (Pkg3) and Pkg REPL-mode because [PkgDev](https://github.com/JuliaLang/PkgDev.jl) doesn't seems to work anymore with Julia 0.7 https://github.com/JuliaLang/PkgDev.jl/issues/143
Creating this dummy package will give you example of all useful files (`.travis.yml`, `appveyor.yml`, `Project.toml` ... ).

Kind regards

**Comment by scls19fr** [comment]
Enabling 
- https://github.com/JuliaComputing/FemtoCleaner.jl (for automatically upgrading deprecated syntax) 
- https://github.com/attobot/attobot (package release bot)
should probably also be considered

PS : I haven't tested attobot since Julia 0.7

**Comment by andrewadare** [comment]
Thanks for these suggestions and your related PR. I enabled Travis CI for the repo.

**Comment by scls19fr** [comment]
So badge URL need to be updated accordingly
See https://github.com/andrewadare/LibSerialPort.jl/pull/21

**Comment by andrewadare** [comment]
The status of this issue:
 - the repo has been moved from my account to JuliaIO
 - Travis was enabled on JuliaIO by @samuelpowell 
 - the README badge has been directed to the new Travis location
 - Travis builds are passing on macOS, but failing on linux due to insufficient port permissions on the build box. PRs #32 and #33 are the configuration attempts.

**Comment by samuelpowell** [comment]
@andrewadare we could ask Travis for debug access to work out what permissions are required (https://docs.travis-ci.com/user/running-build-in-debug-mode/). I'm happy to fire off an email if you think that's a good way forward.

**Comment by andrewadare** [comment]
I appreciate that. I do have to wonder, though, if we're just working to get one step closer to a more fundamental roadblock with this CI approach. If we do achieve permission to connect to /dev/ttyS0 or similar on the Travis build VM, what then? I doubt there's anything at that endpoint. 

The unit tests I wrote were designed to work when connected to something with UART hardware (e.g. a microcontroller running the Arduino sketch in the examples directory). I don't know much about Travis, but I doubt that's an easy thing to emulate. Maybe we should just catch the permissions error, report success as far as the test is concerned (i.e. return nothing), and move on for now. I don't have time to research this further, but in the longer term maybe we can look into something like [AWS Greengrass](https://aws.amazon.com/greengrass/).

**Comment by samuelpowell** [comment]
I quite agree - very difficult to use cloud based CI for hardware stuff. For similar libraries I’ve usually been content that CI successfully installs and loads the library. 

In a related project @ReubenHill recently implemented tests which check different functionality depending if hardware is or is not available (https://github.com/Gowerlabs/LibFTD2XX.jl), this let’s us catch problems in the higher level wrapper. 

For now, and with your permission, I’m happy to make the request to Travis and see if we can at least get to the point where we can open some ports.

**Comment by andrewadare** [comment]
I'd be interested to learn how this Travis configuration is supposed to work, so by all means, feel free to look into it. Of course, you're also welcome to push changes to #33 or supersede it if you come up with something better.

**Comment by samuelpowell** [comment]
Okay I'll see what I can do.

**Comment by samuelpowell** [comment]
@andrewadare Travis have replied but the request must be made by an admin of the repository. Could you give me administrator rights or alternatively send me you email such that I can CC you in? If you want to do the latter the address given [here](https://kclpure.kcl.ac.uk/portal/s.powell.html) is fine.

**Comment by mgkuhn** [comment]
I there still anything in this old ticket that hasn't been resolved yet as of `v0.5.0`?
